ASoC/SoundWire: parse SDCA functions to filter the RT712-VB configuration

plbossart opened the pull request:
Somewhat convoluted code to skip machine driver entries that require RT712 VB hardware.

plbossart committed:
ASoC: sdca: add initial module

Add new module for SDCA (SoundWire Device Class for Audio) support.
For now just add a parser to identify the SDCA revision and the
function mask.

Note that the SDCA definitions and related MIPI DisCo properties are
defined only for ACPI platforms and extracted with _DSD helpers. There
is currently no support for Device Tree in the specification, the
'depends on ACPI' reflects this design limitation. This might change
in a future revision of the specification but for SDCA 1.0 ACPI is the
only supported type of platform firmware.

The SDCA library is defined with static inline fallbacks, which will
allow for unconditional addition of SDCA support in common parts of
the code.

For more technical details, the SDCA specification is available for
public downloads at https://www.mipi.org/mipi-sdca-v1-0-download

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: slave: parse SDCA function mask and revision

Use SDCA helpers to get the information and store it. When platforms
are not based on ACPI the helpers do absolutely nothing.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
@shumingfan @bardliao here's an example of what we could do to filter the required topology.

On a MTL device with RT713, I can detect the VA version.
I don't have a device with RT712 VB but presumably the SmartMic function is the main difference.

plbossart committed:
ASoC: SDCA: add quirk function for RT712_VB match

Add a generic match function for quirks, chances are we are going to
have lots of those...

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: Intel: soc-acpi: add is_device_rt712_vb() helper

Add a filter to skip the RT172 VB configuration if a SmartMic Function
is not found in the SDCA descriptors.

If the ACPI information is incorrect this can only be quirked further
with DMI information.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: SOF: Intel: hda: use machine_quirk() for SoundWire

Use the existing machine_quirk() callback to select an alternate
topology for RT712-VB devices.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
@bardliao @shumingfan I can't figure out if the RT712 and RT713 are identical or not, and if they have the same number of endpoints.

For now I get this error when I align RT712 and RT713...

[  144.018284] sof_sdw sof_sdw: 1 is too many endpoints for codec: 0x713

plbossart committed:
ASoC: Intel: sof_sdw: add SmartMic DAI for RT712 VB

In theory the dailinks are created based on the number of endpoints
reported in ACPI match tables, so it should harmless to add a new
dailink: RT712 VA would not use it since it has only 2 endpoints.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
@shumingfan can you test this with RT712 VA and VB? The topologies required are generated with https://github.com/thesofproject/sof/pull/9125

bardliao commented:
> @bardliao @shumingfan I can't figure out if the RT712 and RT713 are identical or not, and if they have the same number of endpoints.
> 
> For now I get this error when I align RT712 and RT713...
> 
> [ 144.018284] sof_sdw sof_sdw: 1 is too many endpoints for codec: 0x713

rt712 and rt713 are not identical. rt712 supports amp function while rt713 doesn't.

bardliao commented on sound/soc/sdca/sdca_device.c:
```diff
@@ -22,3 +23,26 @@ void sdca_lookup_interface_revision(struct sdw_slave *slave)
+
+	switch (quirk) {
+	case SDCA_QUIRKS_RT712_VB:
+		if ((slave->sdca_interface_revision > 0x0604) &&
```
Maybe >= 0x1000? As far as I know, the function type definition changed from sdca 0.6 to 1.0.

plbossart replied:
@bardliao @shumingfan I could see this value in the ASL file:
Package (2) {"mipi-sdw-sdca-interface-revision", 0x0801}, // Integer: SDCA 0.8r01

Can I use it as an indicator for VB, or just keep the > 0x0604 that was used for VA?

shumingfan replied:
@plbossart Yes, the SDCA version of VB should be version 0x0801 or above.

shumingfan commented:
> @shumingfan can you test this with RT712 VA and VB? The topologies required are generated with [thesofproject/sof#9125](https://github.com/thesofproject/sof/pull/9125)

@plbossart I try to test RT712VA and VB on MTL RVP with different ASL files.
[rt712-l0-disco-table-test.asl.txt](https://github.com/thesofproject/linux/files/15331774/rt712-l0-disco-table-test.asl.txt)
[rt712-vb-l0-disco-table-test.asl.txt](https://github.com/thesofproject/linux/files/15331775/rt712-vb-l0-disco-table-test.asl.txt)
This PR could find out the VA and VB, then load different tplg.

ujfalusi commented on sound/soc/sdca/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config SND_SOC_SDCA
+	tristate "ASoC SDCA library"
+	depends on ACPI
```
Strictly speaking SDCA should not depend on ACPI?

plbossart replied:
All the DisCo stuff is totally dependent on ACPI, and the _DSD properties. No way to use it in Device Tree platforms until a future specification defines how to do it.
There's no way we can remove this dependency for now.

ujfalusi replied:
Right, then in the help it might worth mentioning this?
Do we want COMPILE_TEST coverage enabled?

plbossart replied:
Yeah we can add more explanations.
if ACPI is not defined there are fallback headers with static inlines, so the sdca functions will do nothing.

plbossart replied:
added new comments in the commit message on all this.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -0,0 +1,96 @@
+
+	acpi_dev_for_each_child(adev, find_sdca_function, &slave->sdca_function_mask);
+}
+EXPORT_SYMBOL_NS(sdca_lookup_function_mask, SND_SOC_SDCA);
```
Should this be namespaced to 'acpi' ?

plbossart replied:
no, it's on top of ACPI. The namespace is usually related to the module.

ujfalusi replied:
`sdca_lookup_function_mask()` is hardwired to only work on ACPI because of the `to_acpi_device_node()`, `acpi_dev_for_each_child()` and `acpi_evaluate_integer()`

It is an ACPI function.

plbossart replied:
hence the depends on ACPI. Not sure if I get the feedback here.

ujfalusi commented on sound/soc/codecs/Kconfig:
```diff
@@ -1686,6 +1686,8 @@ config SND_SOC_RT711_SDCA_SDW
 	tristate "Realtek RT712 SDCA Codec - SDW"
 	depends on SOUNDWIRE
+	depends on ACPI
+	select SND_SOC_SDCA
```
you are disabling RT712 for non ACPI architectures, is that right?

plbossart replied:
yes, but in practice there are no users of this device outside of ACPI

we could revisit this and put the select somewhere else. Maybe the machine driver?

ujfalusi replied:
I would handle this ACPI dependency in sound/soc/sdca to compile even if ACPI=n
The machine driver should select ACPI if it is destined to run on ACPI machine

plbossart replied:
machine drivers depends on ACPI.

The problem is really the 'select' part, which overrides the dependencies. So the dependencies have to be moved to the highest level...

ujfalusi replied:
you select SND_SOC_SDCA, SND_SOC_SDCA should not depend on ACPI, if I understand from your other comment, the sdca lib will compile even if no ACPI is selected?

plbossart replied:
no, the sdca lib will not be compiled without ACPI, but the functions have a fallback with static inlines.

That would make e.g. the changes to drivers/soundwire/slave.c where we add those sdca functions completely harmless on non-ACPI platforms.

plbossart replied:
I removed this RT712 change, for now it's not relevant. Only the MTL match needs SDCA in this PR, so I added a select SND_SOC_SDCA to avoid compilation issues reported by the kbuild bot.

ujfalusi commented on sound/soc/intel/common/soc-acpi-intel-mtl-match.c:
```diff
@@ -715,6 +756,30 @@ static const struct snd_soc_acpi_link_adr cs42l43_link0_cs35l56_link2_link3[] =
+	for (i = 0; i < ctx->num_slaves; i++) {
+		if (sdca_device_quirk_match(ctx->ids[i].slave,
+					    SDCA_QUIRKS_RT712_VB))
+			return arg;
```
is this right?
`arg` is `struct sdw_intel_ctx *` and now we return it as `struct snd_soc_acpi_mach *` ?

Yes, I have read the comment, still this isn't right

plbossart replied:
The only way to do this right would be to add a new callback. Not sure it's more elegant really.

ujfalusi replied:
it is going to be a matter of time when someone interprets the returned pointer assuming it is pointing to a `struct snd_soc_acpi_mach`.

plbossart replied:
I know, I am just not hot on adding a 'machine_quirk_match' callback in the ACPI tables just for the SoundWire case.

plbossart replied:
added a 'machine_check' callback.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1202,7 +1202,8 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
 		if (i == hdev->info.count || !link->num_adr)
-			break;
+			if (!mach->machine_quirk || mach->machine_quirk(hdev->sdw))
+				break;
```
I'm not sure how to put this now into a sentence, how to interpret it? Why we break out if `machine_quirk()` returns non 0?

plbossart replied:
if the machine quirk results a non-NULL value, that means the machine is valid.
If the machine quirk returns NULL, we want to keep looking for an alternate entry in the table.

You have to look at the table, we have the VB first then VA. If we have all the peripherals listed with the expected _ADR, the machine quirk will either stop on the VB entry or skip to use VA.

ujfalusi replied:
I mean, in context of the level up comment..
We have found a match if I read it right and now we are going to run an 'are you sure' query on it?

plbossart replied:
yes, the initial match is really only a partial match based on _ADR information, but we need a second level filter to disambiguate identical configurations used for different types of hardware.
If this looks like a mess it's because it is. I've share suggested text to be added to the SDCA spec to prevent this from happening, but for now it is what it is.

plbossart replied:
I added a "machine_check" callback to solve this return value problem

ujfalusi reviewed (changes requested):
@plbossart, we should not lock SDCA to be only usable with ACPI, that sounds wrong.
I find some other parts confusing as well, probably documentation can help address those.

plbossart commented:
Of course we have a side effect with MTL AIOC having a bad DSDT

````
[    6.321319] kernel: acpi device:2e: found invalid SDCA function AF01 type 4 ADR 1, skipped
[    6.321847] kernel: acpi device:35: found invalid SDCA function AF02 type 2 ADR 2, skipped
````

Gah. what a mess.

plbossart committed:
ASoC: soc-acpi: introduce new 'machine check' callback

The existing machine_quirk() returns a pointer to a soc_acpi_mach
structure.
For SoundWire/SDCA support, we need a slightly different
functionality where a quirk function either validates or NACKs an
initial selection, based on additional firmware/DMI information.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented on sound/soc/sdca/sdca_device.c:
```diff
@@ -22,3 +23,30 @@ void sdca_lookup_interface_revision(struct sdw_slave *slave)
+		    (slave->sdca_function_mask &
+		     BIT(SDCA_FUNCTION_TYPE_SMART_MIC)) &&
+		    (id->mfg_id == 0x025d) &&
+		    ((id->part_id == 0x712) ||
```
@shumingfan should I be more strict and only use the partID 0x712? I don't think we should use the 0x713 since it's a different part, not sure on 0x716 and 0x717 (if they exist?)

shumingfan replied:
The version B also includes RT713VB/716VB/717VB.

plbossart replied:
sorry @shumingfan what is RT713VB? Does it have a SmartMic integrated as well, i.e. the RT1713 device is no longer exposed on the bus?

I am so lost here.

shumingfan replied:
> Does it have a SmartMic integrated as well, i.e. the RT1713 device is no longer exposed on the bus?

Yes, VB has only one SoundWire interface which supports multi-lanes and SmartMic. Hence, there is no longer with RT1713 in the version B.

plbossart commented:
https://sof-ci.01.org/linuxpr/PR4995/build3008/devicetest/index.html?model=MTLP_SDW_AIOC&testcase=verify-kernel-boot-log is a case of bad configuration I am afraid on ba-mtlp-sdw-aioc-02: the device is detected as HDaudio and we use it for SoundWire tests??

@ssavati can you please double-check this device? This is clearly wrong:
````
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof_intel_hda_generic:hda_init_caps: sof-audio-pci-intel-mtl 0000:00:1f.3: skipping SoundWire, no links enabled
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof_intel_hda:hda_codec_probe: sof-audio-pci-intel-mtl 0000:00:1f.3: HDA codec #2 probed OK: response: 8086281d
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof_intel_hda:request_codec_module: hdaudio ehdaudio0D2: loading codec module: hdaudio:v8086281Dr00100000a01
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: hda codecs found, mask 4
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: using HDA machine driver skl_hda_dsp_generic now
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_intel_dspcfg:intel_nhlt_get_dmic_geo: sof-audio-pci-intel-mtl 0000:00:1f.3: found 1 format definitions
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_intel_dspcfg:intel_nhlt_get_dmic_geo: sof-audio-pci-intel-mtl 0000:00:1f.3: max channels found 2
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_intel_dspcfg:intel_nhlt_get_dmic_geo: sof-audio-pci-intel-mtl 0000:00:1f.3: Array with 2 dmics
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_intel_dspcfg:intel_nhlt_get_dmic_geo: sof-audio-pci-intel-mtl 0000:00:1f.3: dmic number 2 max_ch 2
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-mtl 0000:00:1f.3: Module parameter used, changed fw path to intel/sof-ipc4/mtl/community
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-mtl 0000:00:1f.3: Module parameter used, changed tplg path to intel/sof-ipc4-tplg
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-mtl 0000:00:1f.3: Module parameter used, changed tplg name to sof-mtl-rt711-l0-rt1316-l23-rt714-l1.tplg
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Firmware paths/files for ipc type 1:
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file:     intel/sof-ipc4/mtl/community/sof-mtl.ri
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware lib path: intel/sof-ipc4-lib/mtl
May 16 18:08:04 ba-mtlp-sdw-aioc-02 kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file:     intel/sof-ipc4-tplg/sof-mtl-rt711-l0-rt1316-l23-rt714-l1.tplg
````

ssavati commented:
@plbossart I have checked the BIOS configuration and other settings , everything is fine. Same device without this PR works fine. I think there is regression in this PR.

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -489,6 +489,7 @@ struct sdw_slave_id {
 struct sdw_extended_slave_id {
+	struct sdw_slave *slave;
 	int link_id;
 	struct sdw_slave_id id;
```
The `id` becomes redundant, we should remove it later (not a scope of the PR).

plbossart replied:
yes, that's true, maybe we can do better. let me think about this.

plbossart replied:
done

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -0,0 +1,96 @@
+	return 0;
+}
+
+void sdca_lookup_function_mask(struct sdw_slave *slave)
```
What I meant is that we might want to have `sdca_acpi_lookup_function_mask()` or use internally:
```
struct acpi_device *adev;

/* Only ACPI based lookup is supported */
if (!is_acpi_node(slave->dev.fwnode))
	return;

adev = to_acpi_device_node(slave->dev.fwnode);
```
to make the function generic and open ended for non acpi platforms?

plbossart replied:
All SDCA platforms are based on ACPI so far.
For example the 'sdca-initialization-table' is defined as an ACPI "Buffer". 

We can revisit this when the MIPI WG issues a specification which supports Device Tree. That may take a while.

Note that if ACPI is not defined then the static inline fallback functions do nothing, which is similar to what you suggested in the test above.

plbossart replied:
no change, I can't test the case without ACPI anyways.

ujfalusi commented on drivers/soundwire/slave.c:
```diff
@@ -87,6 +88,8 @@ int sdw_slave_add(struct sdw_bus *bus,
 	}
 	sdw_slave_debugfs_init(slave);
 
+	sdca_lookup_function_mask(slave);
```
OK, definitely the `is_acpi_node()` way would work in this generic path.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -183,6 +183,9 @@ struct snd_soc_acpi_link_adr {
  * audio codecs whose presence if checked with ACPI
+ * @machine_check: pointer to quirk function. The functionality is similar to
+ * @machine_quirk, except that the return value is a boolean: the intent is
+ * to skip a machine if the additional hardware/firmware verification
```
The order of machine_check/machine_quirk is not aligned withe the struct position.

Also it looks like the sentence is just cut in half?

plbossart replied:
fixed.

ujfalusi commented on sound/soc/intel/common/soc-acpi-intel-mtl-match.c:
```diff
@@ -715,6 +756,30 @@ static const struct snd_soc_acpi_link_adr cs42l43_link0_cs35l56_link2_link3[] =
+ * and return types are not the traditional 'struct snd_soc_acpi_mach'
+ * pointer but instead the sdw_intel_ctx which contains the peripheral
+ * information required for the SoundWire/SDCA filter on the SMART_MIC
+ * setup and interface revision.
```
The comment is now outdated.

plbossart replied:
thanks for noticing this, fixed.

ujfalusi reviewed (commented):
@plbossart, It looks much cleaner, I have few suggestions around the acpi handling to make the code more generic.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -200,6 +203,7 @@ struct snd_soc_acpi_mach {
 	const char *tplg_filename;
 	const char *board;
 	struct snd_soc_acpi_mach * (*machine_quirk)(void *arg);
+	bool (*machine_check)(void *arg);
```
probably it is going to be too much overload, but the functionality of this new callback is `is_machine_match()`

ujfalusi commented on sound/soc/sdca/sdca_device.c:
```diff
@@ -22,3 +23,30 @@ void sdca_lookup_interface_revision(struct sdw_slave *slave)
 }
 EXPORT_SYMBOL_NS(sdca_lookup_interface_revision, SND_SOC_SDCA);
+
+bool sdca_device_quirk_match(struct sdw_slave *slave, enum sdca_quirk quirk)
```
In my reading a 'quirk' does something active, while this function is more like match checking to sort codecs out from different parts or something?

If you would have `SDCA_QUIRKS_RT712_VA` then it would return true if the codec matches with that family.

I'm not sure if quirk is the right term for this.

plbossart replied:
it's what we use for DMI quirks. If we have  a match, we do something.

plbossart replied:
and 'quirk' is the term we use when we have hardware than requires additional information to be special-cased.

plbossart committed:
ASoC/soundwire: remove sdw_slave_extended_id

This structure is used to copy information from the 'sdw_slave'
structures, it's better to create a flexible array of 'sdw_slave'
pointers and directly access the information. This will also help
access additional information stored in the 'sdw_slave' structure,
such as an SDCA context.

This patch does not add new functionality, it only modified how the
information is retrieved.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
> @plbossart I have checked the BIOS configuration and other settings , everything is fine. Same device without this PR works fine. I think there is regression in this PR.

Thanks for checking @ssavati there was a one-line Kconfig change that caused the problem by completely disabling SoundWire with the default kconfig scripts. Fixed now on my side.

plbossart commented:
New version with a cleaner parsing of the functions with the backwards-compatible re-ordering for early devices
````
[  131.499758] soundwire sdw:0:0:025d:0713:01: found SDCA function UAJ (type 6)
[  131.499761] soundwire sdw:0:0:025d:0713:01: found SDCA function HID (type 10)
[  131.500270] soundwire sdw:0:1:025d:1316:01: found SDCA function SmartAmp (type 1)
[  131.500804] soundwire sdw:0:2:025d:1316:01: found SDCA function SmartAmp (type 1)
````

@shumingfan please re-check if I did something wrong, it's Friday afternoon and it's been a long week. Thanks!

marc-hb commented:
https://github.com/thesofproject/linux/actions/runs/9134952643/job/25121470472?pr=4995
```
sound/soc/sof/amd/acp-common.c: In function ‘amd_sof_sdw_machine_select’:
sound/soc/sof/amd/acp-common.c:148:86: error: ‘struct sdw_amd_ctx’ has no member named ‘ids’
  148 |                                                                         acp_data->sdw->ids,
      |                                                                                      ^~
sound/soc/sof/amd/acp-common.c:149:86: error: ‘struct sdw_amd_ctx’ has no member named ‘num_slaves’
  149 |                                                                         acp_data->sdw->num_slaves))
      |                                                                                      ^~
sound/soc/sof/amd/acp-common.c:147:38: error: too many arguments to function ‘snd_soc_acpi_sdw_link_slaves_found’
  147 |                                 if (!snd_soc_acpi_sdw_link_slaves_found(sdev->dev, link,
```

bardliao reviewed (dismissed):
LGTM

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -0,0 +1,153 @@
+		case 6: function_mask_patch |= BIT(SDCA_FUNCTION_TYPE_HID);
+			break;
+		default:
+			dev_warn(&slave->dev, "%s: found unsupported early draft SDCA function type %d, skipped\n",
```
nitpick: you can use the `dev` for the print and the line is still too long.

plbossart replied:
fixed, thanks

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -0,0 +1,153 @@
+				__func__, function_topology);
+			continue;
+		}
+		dev_info(&slave->dev, "found SDCA function %s (type %d)\n",
```
Use `dev` only for the prints consistently.

plbossart replied:
fixed.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -0,0 +1,153 @@
+}
+EXPORT_SYMBOL_NS(sdca_lookup_function_mask, SND_SOC_SDCA);
+
+MODULE_LICENSE("Dual BSD/GPL");
```
does it needs to be dual license?

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -201,6 +205,7 @@ struct snd_soc_acpi_mach {
 	const char *board;
 	struct snd_soc_acpi_mach * (*machine_quirk)(void *arg);
 	const void *quirk_data;
+	bool (*machine_check)(void *arg);
```
this still sounds strange for first and second read, but the `machine_quirk` which returns `snd_soc_acpi_mach` pretty much covers other options (like `machine_match`), I guess it is just fine.

ujfalusi reviewed (commented):
@plbossart, thank you for the updates, looks good with the exception of the mixed slave->dev / dev use for prints.

plbossart commented on include/sound/sdca.h:
```diff
@@ -0,0 +1,37 @@
+ */
+struct sdca_device_data {
+	u32 interface_revision;
+	u32 function_mask;
```
Note to self: this function_mask assumed that there is only ONE type of function type per codec. The spec says otherwise, it's permitted to have multiple instances of the same function type, except for HID. That means we need an array of (ADR, function_type) pairs.

plbossart reviewed (commented):
the definition of a function_mask was a bad idea...

plbossart commented:
closing for now, WIP alternate solution with extensions for functions
